libs/execution_context/ebpf_native.c

```diff
@@ -1790,29 +1796,35 @@ ebpf_native_load_programs(
         ebpf_lock_unlock(&_ebpf_native_client_table_lock, state);
         lock_acquired = false;
     }
-    if (result != EBPF_SUCCESS) {
-        if (maps_created) {
-            _ebpf_native_clean_up_maps(instance.maps, instance.map_count, true, true);
-            instance.maps = NULL;
-            instance.map_count = 0;
-        }
 
-        if (programs_created) {
-            _ebpf_native_clean_up_programs(instance.programs, instance.program_count, true);
-            instance.programs = NULL;
-            instance.program_count = 0;
+    if (programs_created) {
+        for (size_t i = 0; i < instance.program_count; i++) {
+            _ebpf_native_clean_up_program_handle(instance.programs[i]);
+            // Free native program context only if the program is not loaded. If the program
+            // is loaded, it will be freed when the ebpf program object is freed.
+            if (!instance.programs[i]->loaded) {
+                _ebpf_native_clean_up_program(instance.programs[i]);
+            }
+            instance.programs[i] = NULL;
         }
+    }
+    ebpf_free(instance.programs);
+    instance.programs = NULL;
+    instance.program_count = 0;
 
-        ebpf_free(local_service_name);
-    } else {
-        // Success case. No need to close program and map handles.
-        // Free the map contexts.
-        _ebpf_native_clean_up_maps(instance.maps, instance.map_count, false, false);
+    if (maps_created) {
+        if (result != EBPF_SUCCESS) {
+            _ebpf_native_clean_up_maps(instance.maps, instance.map_count, true, true);
```

Comment:
for readability some comment on the two boolean parameters or using descriptive constants instead of literal would help.